### ISSUE TITLE
UnusedValueConstructor false positive?

### ISSUE THREAD

**Issue by Augustin82** [stil4m/elm-analyse#215]
Hi there,

I have the following in my app:
```
type Page
    = LandingPage
    | FormPage
    | ConfirmationPage POId
```

And later in the same file, I pattern match on it:
```    case model.page of
        LandingPage ->
            -- do stuff

        FormPage ->
            -- do stuff

        ConfirmationPage poId ->
            -- do stuff
```
But I get an elm-analyse error on the Type declaration, stating:
```[elm-analyse 272] [W] Value constructor `ConfirmationPage` is not used. Declared 
See https://stil4m.github.io/elm-analyse/#/messages/UnusedValueConstructor```

Am I missing something obvious here? Is the problem with my toolchain? I'm using nvim, Coc and elm-language-server.

Thanks!

**Comment by jfmengels** [comment]
The problem is that the `ConfirmationPage` constructor is never **used**. Meaning, yes, you handle it in pattern matching expressions where it should be handled, but you never **construct** this value anywhere in your code.
This rule aims to show you that you could remove the *ConfirmationPage* declaration.

**Comment by Augustin82** [comment]
So... Maybe I'm dense, but how could I pattern match on that value if I
didn't declare it there?

**Comment by jfmengels** [comment]
Let's take the following example (imagine it's the whole file):

```elm
module A exposing (SomeType, doSomething, init)

type SomeType = A | B

doSomething : SomeType -> String
doSomething someType =
  case someType of
    A -> "a"
    B -> "b"

init : SomeType
init = A
```

Here, the rule would report that B is never used. What it means is, that there is never a case where a value with sub-type B gets constructed (no function here creates a value of type SomeType using the B constructor), so handling the case for B in the pattern matching expressions is unnecessary (as in code that will never be executed).

In practice, what the rule urges you to do, is to remove B from SomeType (ConfirmationPage from Page) and have the code simplified to the following:

```elm
module A exposing (SomeType, doSomething, init)

type SomeType = A

doSomething : SomeType -> String
doSomething someType =
  case someType of
    A -> "a"

init : SomeType
init = A
```

**Comment by Augustin82** [comment]
OK, confirming the "I'm dense" hypothesis. I completely missed that point in your first message, and yeah you (and elm-analyse) are obviously right.

That value is indeed a leftover from some refactoring, and I just didn't question it, hence my missing the point.

Closing this now, many thanks to you @jfmengels!
